source/storage/storage.go

```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"fmt"
+	"io/ioutil"
+)
+
+// Source implements FeatureSource.
+type Source struct{}
+
+// Name returns an identifier string for this feature source.
+func (s Source) Name() string { return "storage" }
+
+// Discover returns feature names for storage: nonrotational if any SSD drive present.
+func (s Source) Discover() ([]string, error) {
+	features := []string{}
+
+	// Check is there zero in any device rotational status
+	blockdevices, err := ioutil.ReadDir("/sys/block/")
+	if err == nil {
+		for _, bdev := range blockdevices {
+			fname := "/sys/block/" + bdev.Name() + "/queue/rotational"
+			bytes, err := ioutil.ReadFile(fname)
+			if err != nil {
+				return nil, fmt.Errorf("can't read rotational status: %s", err.Error())
+			}
+			if bytes[0] == byte('0') {
```

Comment:
Do you know if this is the standard way of figure out whether a disk is non-rotational or not? Are there any other ways to figure it out?

Reply:
Will this work in both a VM and bare-metal environment?

Reply:
Reading /sys/block/DEV/queue/rotational is AFAIK the only universal way without using more specialized disk tools which may be specific to technology or vendor. Some sources recommend lsblk but this utility simply reads same location under /sys/block. If someone knows better or more "standard" way, please speak up.

About VM: situation is not very good from detection POV. If a VM runs on top of bare metal where non-rotational disk is used, VM likely cant detect it easily, as (at least typically I think?) disk device is virtualized for VM, and "non-rotational" status gets lost in VM even if phys device has it. It is possible that some virtualization methods allow exposing device directly, then situation can be better.
In similar fashion, non-rotational capability gets hidden/lost by RAID controller if RAID controller presents virtual storage to host.

Reply:
Thanks for looking at this.